### PR TITLE
Reduce duplication in views & reselect for state management

### LINKED ISSUE #48: Reduce duplication in handlers & views

### PR DESCRIPTION
**Pull request by katbow**
Closes #48, #91, #67

+ We've made all of the views handlebars partials now (plus other components as partials too).
+ We've added in reselect to organise our state object

### CODE REVIEW

File: public/src/js/registerPartials.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+var Handlebars = require('handlebars/runtime').default;
+var hbsPartials = {
```
Comment (andrewMacmurray) [inline public/src/js/registerPartials.js]: we're aware that there's LOTS of duplication here but the scripts were breaking when we loaded them in a forEach or a map, (may need to have a look at loading them in a different way)

File: public/views/actionplan.html [hunk 2]
```diff
@@ -3,55 +3,12 @@
   <head>
```
Comment (andrewMacmurray) [inline public/views/actionplan.html]: also lots of duplication with the html files (the only thing different between them is the hbs target), but we were struggling how to detect which page was being loaded to render it with the correct partial. Any thoughts here would be welcome 😄

File: public/src/js/selectors.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
+const { createSelector } = require('reselect');
+const merge = require('./helpers.js').utils.merge;
+
+const _currentClient = state => state.currentClient;
```
Comment (andrewMacmurray) [inline public/src/js/selectors.js]: naming the selectors with the word `selector` was getting really long! (`currentClientObjectSelector` lol). Not sure about the underscore though, any suggestions?

File: public/views/actionplan.html [hunk 4]
```diff
@@ -3,55 +3,12 @@
...
-              {{else}}
-              <span>Companion</span>
-              {{/if}}
-            </div>
-          </div>
-        {{/each}}
-      </script>
-    </div>
-    <script src="/src/js/vendor/handlebars.min.js" charset="utf-8"></script>
+    <div hbs_target_actionplanview></div>
```
Comment (andrewMacmurray) [inline public/views/actionplan.html]: also lots of duplication with the html files (the only thing different between them is the hbs target), but we were struggling how to detect which page was being loaded to render it with the correct partial. Any thoughts here would be welcome 😄

**Comment by rjmk** [comment]
I'm having some trouble leaving line comments this evening! Looks very good; I'm very happy with the partials (though I find requiring 'html' files a bit confusing).

On `selectors.js`:

I might neaten [this](https://github.com/ageuk/companion-app/blob/8b0570292126bb6fdcb498febccb51277a6bef26/public/src/js/selectors.js#L4-L11) up as something like

```js
var baseSelectors = {}
var baseProps = [ 'currentClient', 'currentTopic', ...]

baseProps.forEach(prop => baseSelectors[prop] = state => state[prop])
```

But that's mostly personal taste. By the way, are we able to use arrow functions on the frontend now?

Another comment on the same bit of code: does the difference in case between `state.currentClient` and `state.client_actions` indicate something in specific?

Only other comment is on `transformState` -- it's not necessary to have one big selector that transforms the state in one go. It's perfectly fine to do so if you prefer, though! Just thought I'd point out many use reselect purely to create a bunch of getter functions used throughout the code.  
My only concern with `transformState` is to remember that you must be careful with the output of that state -- you don't want to be transforming it. You only want to change the original state and make sure that you're always plugging that through `transformState` when it may have been updated.


Anyway, nothing that has to be changed on this PR. A couple of tweaks to make if you fancy, and a couple of questions to check understanding. I'll assign to you. Let me know if you have any questions, make any changes you want to make, then assign back to me

**Comment by andrewMacmurray** [comment]
Thanks Rafe! @katbow and I are back in on Monday so will have a look over it then :smile:

**Comment by andrewMacmurray** [comment]
@rjmk the difference between `currentClient` etc and `client_actions` is that `currentClient` is a piece of ui state but `client_actions` is data that is going back up to the db, not really sure if there's a good way to indicate this difference

**Comment by andrewMacmurray** [comment]
@rjmk as far as using arrow functions on the frontend, no we can't yet 😞 , we just wrote it out like that at first as it was easier for us to understand, is it worth using something like babel or is that overkill?

**Comment by andrewMacmurray** [comment]
@rjmk I think we'll probably keep the one big selector, the reason is the sql data is coming in one big object so we'd have to split it up to perform all the different transforms on it, is there any way we can mark transform state so it's not misused (I know this isn't ideal but at least some way of protecting it)

File: public/src/js/selectors.js [hunk 5]
```diff
@@ -1,37 +1,45 @@
...
+  'client_actions'
+];
 
-const _actions = state => state.actions;
-const _prompts = state => state.prompts;
-const _clients = state => state.clients;
-const _clientActions = state => state.client_actions;
+const baseSelectors = baseProps
+  .map(prop => state => state[prop])
+  .reduce((a, b, i) => {
```
Comment (rjmk) [inline public/src/js/selectors.js]: Happy with the `reduce`, but note it's just the same as a `forEach`, as you're still mutating an object

**Comment by rjmk** [comment]
> is it worth using something like babel or is that overkill?

**Comment by rjmk** [comment]
>I think we'll probably keep the one big selector, the reason is the sql data is coming in one big object so we'd have to split it up to perform all the different transforms on it, is there any way we can mark transform state so it's not misused (I know this isn't ideal but at least some way of protecting it)

I don't think there is. We could freeze it, but we're mostly avoiding mutation. I'm not sure I understand the point about the large object having to be split up. Could you given an example of what you mean?

File: public/src/js/selectors.js (hunk 5 above)
Comment (andrewMacmurray) [inline public/src/js/selectors.js]: whoops, yea the linter was complaining about assignment in a function, will change it to object assign

Comment (rjmk) [inline public/src/js/selectors.js]: OK, though the mutation is pretty harmless in this case

**Comment by andrewMacmurray** [comment]
when we're getting the db object [here](https://github.com/ageuk/companion-app/blob/33d1bf792e9c3e776fa9c795662a9a5dbf1749e0/public/src/js/main.js#L10), we'd have to either mutate each of the keys or make a new object out of each of them and combine them back together, or change what data we're passing into the partials, it just seems like a lot of rewiring

**Comment by rjmk** [comment]
@andrewMacmurray OK. I think I see. The way it would probably be written without the big selector is to have `db.getData` use the reselect functions. That way, you can keep the localStorage state flat.

**Comment by andrewMacmurray** [comment]
@rjmk ah that's a good idea, will try that out

**Comment by andrewMacmurray** [comment]
@rjmk I've moved the selectors into main.js where the `db.getData('db')` call is, I'm not really sure how that changes things but do you think that reads any clearer?

**Comment by rjmk** [comment]
@andrewMacmurray I was thinking more of the `helpers.js` `db.getData` function.

It might look something like

```js
db.getData = function (key) {
  var selector = dataSelectors[key]
  return utils.safeParse(selector(data))
```

(We might consider moving the `safeParse` to the selector functions as well)

**Comment by andrewMacmurray** [comment]
@rjmk But we've got all the localStorage data as one big object, and each of those getData calls would depend on other keys in the bog data object, I think that approach would require a lot of restructuring.

**Comment by rjmk** [comment]
@andrewMacmurray Sorry, when I wrote `utils.safeParse(selector(data))` I should have written `utils.safeParse(selector(localStorage[key]))`. Does that have the same problem do you think?

(Except for the fact that the `safeParse` should almost certainly be moved into the selectors in this case)

**Comment by andrewMacmurray** [comment]
I think its the same problem really,

`db.getData('db')` returns an object like:

```js
{
  clients: [...],
  topics: [...],
  prompts: [...],
  actions: [...].
  client_actions: [...]
}
```

I think we might be a bit lost in translation 😆  are you in any time this week? Would be good to talk it through

**Comment by rjmk** [comment]
I wrote a response to this, but I think I need to reflect on it more. Bother me tomorrow morning if I've not replied

**Comment by rjmk** [comment]
I'm going to merge this. Here are some thoughts for future discussion.

We've hewed a bit too tightly to code that was ultimately only introduced as placeholder code, I think. Specifically, the `dbHelpers`. My concern can be viewed as one of code separation. The querying of the data should really happen in its own file -- not in a `main` where other stuff might happen. The neat data should never be accidentally be manipulated.

If this were more of an 'app', with dynamic rerendering of the UI based off the state, I think it would be important to pick of the desired bits of state and feed them to the given views. It still might make sense to register the selectors as Handlebars helpers and have them used in the views -- this would be excellent for avoiding accidental updates to the derived state rather than the core state.

This is also tied up with how we integrate state and DOM updates. It may not be desirable to fully rerender the handlebars template with new data, so we will probably need to use events to update the state and the DOM in lockstep

**Comment by andrewMacmurray** [comment]
@rjmk ok that's great, thanks.

That separation of concerns makes a lot more sense to me (separating the db calls from the manipulation), we should definitely have a talk though about the best way to do this just so we're clear on where to go with it

**Pull request merged**